src/utils/image-manager.ts

```diff
@@ -0,0 +1,312 @@
+	return major >= 2018; // note: (NaN >= 2018) is false
+};
+
+/**
+ * @summary Get the most recent compatible version
+ * @function
+ * @protected
+ *
+ * @param {String} deviceType - device type slug or alias
+ * @param {String} versionOrRange - supports the same version options
+ * as `balena.models.os.getMaxSatisfyingVersion`.
+ * See `manager.get` for the detailed explanation.
+ * @returns {Promise<String>} the most recent compatible version.
+ */
+const resolveVersion = async (deviceType: string, versionOrRange: string) => {
+	const balena = getBalenaSdk();
+	const version = await balena.models.os.getMaxSatisfyingVersion(
+		deviceType,
+		versionOrRange,
+		isESR(versionOrRange) ? 'esr' : 'default',
+	);
+	if (!version) {
+		throw new Error('No such version for the device type');
+	}
+	return version;
+};
+
+/**
+ * @summary Get an image from the cache
+ * @function
+ * @protected
+ *
+ * @param {String} deviceType - device type slug or alias
+ * @param {String} version - the exact balenaOS version number
+ * @returns {Promise<fs.ReadStream>} image readable stream
+ *
+ * @example
+ * utils.getImage('raspberry-pi', '1.2.3').then (stream) ->
+ * 	stream.pipe(fs.createWriteStream('foo/bar.img'))
+ */
+const getImage = async (deviceType: string, version: string) => {
+	const imagePath = await getImagePath(deviceType, version);
+	const fs = await import('fs');
+	const stream = fs.createReadStream(imagePath) as ReturnType<
+		typeof fs.createReadStream
+	> & { mime: string };
+	// Default to application/octet-stream if we could not find a more specific mime type
+
+	const { getType } = await import('mime');
+	stream.mime = getType(imagePath) ?? 'application/octet-stream';
+	return stream;
+};
+
+/**
+ * @summary Get a writable stream for an image in the cache
+ * @function
+ * @protected
+ *
+ * @param {String} deviceType - device type slug or alias
+ * @param {String} version - the exact balenaOS version number
+ * @returns {Promise<fs.WriteStream & { persistCache: () => Promise<void>, removeCache: () => Promise<void> }>} image writable stream
+ *
+ * @example
+ * utils.getImageWritableStream('raspberry-pi', '1.2.3').then (stream) ->
+ * 	fs.createReadStream('foo/bar').pipe(stream)
+ */
+const getImageWritableStream = async (deviceType: string, version?: string) => {
+	const imagePath = await getImagePath(deviceType, version);
+
+	// Ensure the cache directory exists, to prevent
+	// ENOENT errors when trying to write to it.
+	const path = await import('path');
+	const { mkdirp } = await import('mkdirp');
+	await mkdirp(path.dirname(imagePath));
+
+	// Append .inprogress to streams, move them to the right location only on success
+	const inProgressPath = imagePath + '.inprogress';
+	const { promises, createWriteStream } = await import('fs');
+	const stream = createWriteStream(inProgressPath) as ReturnType<
+		typeof createWriteStream
+	> &
+		Record<'persistCache' | 'removeCache', () => Promise<void>>;
```

Comment:
Would appreciate typings this explicitly since I think it justifies a bit more why we whose those pros:
```suggestion
	type ImageWritableStream = ReturnType<typeof createWriteStream> &
		Record<'persistCache' | 'removeCache', () => void>;
	const stream = createWriteStream(inProgressPath) as ImageWritableStream;
```
This way you can also add `Promise<ImageWritableStream>` as the return type of the function :)